src/cunumeric/matrix/diag_template.inl

```diff
@@ -14,66 +14,97 @@
  *
  */
 
+#include "cunumeric/pitches.h"
+
 namespace cunumeric {
 
 using namespace Legion;
 using namespace legate;
 
-template <VariantKind KIND, LegateTypeCode CODE>
+template <VariantKind KIND, LegateTypeCode CODE, int DIM, bool extract>
 struct DiagImplBody;
 
 template <VariantKind KIND>
 struct DiagImpl {
-  template <LegateTypeCode CODE>
+  template <LegateTypeCode CODE, int DIM>
   void operator()(DiagArgs& args) const
   {
     using VAL = legate_type_of<CODE>;
 
-    auto shape = args.matrix.shape<2>();
+    if (args.extract) {
+      auto shape_in = args.matrix.shape<DIM>();
+      Pitches<DIM - 1> pitches_in;
+      size_t volume_in = pitches_in.flatten(shape_in);
+      if (volume_in == 0) return;
+      auto shape_out        = args.diag.shape<DIM>();
+      size_t diag_start_dim = DIM - args.naxes;
+      coord_t start         = shape_in.lo[diag_start_dim];
+      coord_t end           = shape_in.hi[diag_start_dim];
 
-    // Solve for the start
-    // y = x
-    // x >= shape.lo[0]
-    const Point<2> start1(shape.lo[0], shape.lo[0]);
-    // y >= shape.lo[1]
-    const Point<2> start2(shape.lo[1], shape.lo[1]);
-    // See if our shape intersects with the diagonal
-    if (!shape.contains(start1) && !shape.contains(start2)) return;
+      for (int i = diag_start_dim + 1; i < DIM; i++) {
+        start = std::max(start, shape_in.lo[i]);
+        end   = std::min(end, shape_in.hi[i]);
+      }
+      coord_t distance = end - start + 1;
+      if (distance < 0) distance = 0;
```

Comment:
Can you just `if (distance <= 0) return;` ?

Reply:
Yes, thanks

Reply:
Actually, if I do this I get an error:

```
Found an uninitialized Legate store
12119
legion_python: core/data/store.cc:125: legate::ReturnValue legate::FutureWrapper::pack() const: Assertion `false' failed.
12120
Signal 6 received by node 0, process 54795 (thread 7f1578f3c000) - obtaining backtrace
12121
Signal 6 received by process 54795 (thread 7f1578f3c000) at: stack trace: 17 frames
12122
  [0] = /usr/lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7f1585372210]
12123
  [1] = /usr/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb) [0x7f158537218b]
12124
  [2] = /usr/lib/x86_64-linux-gnu/libc.so.6(abort+0x12b) [0x7f1585351859]
12125
  [3] = /usr/lib/x86_64-linux-gnu/libc.so.6(+0x25729) [0x7f1585351729]
12126
  [4] = /usr/lib/x86_64-linux-gnu/libc.so.6(+0x36f36) [0x7f1585362f36]
12127
  [5] = /legate/legate.core-install/lib/liblgcore.so(+0x3fde2) [0x7f153f41bde2]
12128
  [6] = /legate/legate.core-install/lib/liblgcore.so(legate::TaskContext::pack_return_values() const+0xae) [0x7f153f43e5be]
12129
  [7] = /legate/legate.core-install/lib/libcunumeric.so(void Legion::LegionTaskWrapper::legion_task_wrapper<legate::ReturnValues, &(legate::ReturnValues legate::LegateTask<cunumeric::DiagTask>::legate_task_wrapper<&cunumeric::DiagTask::gpu_variant>(Legion::Task const*, std::vector<Legion::PhysicalRegion, std::allocator<Legion::PhysicalRegion> > const&, Legion::Internal::TaskContext*, Legion::Runtime*))>(void const*, unsigned long, void const*, unsigned long, Realm::Processor)+0xbc) [0x7f12801f68dc]
  ```

Reply:
that means that your task didn't write to some of the legate stores you passed with reduction or write permission

Reply:
I see. I will try to add the check for the case when `diagonal==0 ` on the python side so this condition is never true on the C++ side.

Reply:
Actually, I can't prevent `diagonal <=0 ` on the C++ side since this is a diagonal for a particular partition. 
But moving `if (distance <= 0) return;` after I get accessor for the reduction instance seems to fix the error.

Reply:
FYI, creating an accessor on a scalar store creates a temporary buffer to hold the update that is later packed into the task's future. Though that is enough to make you assertion go away, the buffer doesn't necessarily have a correct value. The only reason that your code worked is because the reduction accessor also initializes the buffer with the identity: https://github.com/nv-legate/legate.core/blob/branch-22.01/src/core/data/store.inl#L174-L191

In general, passing a store with either write or reduction permission and not actually updating it is dangerous, so you need to make sure that that store actually gets any update no matter what. The assertion you hit was there as a safe guard.

Reply:
In this particular case, the data we reduce is a "promoted" array. To my understanding, when "promoted" array is partitioned to be passed to the task, some index points will have partitions of the "promoted" array that are actually pointing to the same data. In this case, would it be safe if only one index point modify the data?

Reply:
The requirement for a store with reduction permission, however it was constructed, is that tasks provide reduction contributions for a portion of the store and the untouched part will contain the identity value. The requirement for a store with write permission, again however it was constructed, is that tasks should update the entire store, otherwise the untouched part will contain undefined values. The same is true for Legion tasks in general, so there's nothing special. Tasks shouldn't be aware of how the stores were transformed, unless they have to pass the raw pointers to a third-party library.

Coming back to your case, whether the array is promoted or not, it's perfectly fine for only one point task to provide reduction contributions for the array, because, like I said above, the rest will be initialized with the identity. The previous comment was exactly this; the reason your fix worked is because an "uninitialized" reduction store is filled with the identity. (In fact, I think the code should have worked even when you didn't create an accessor, if this was a reduction store. I opened an issue for this.)

Reply:
I see. Thank you for the explanations!